src/module/canvas/perception/modes.ts

```diff
@@ -152,12 +152,57 @@ class DetectionModeTremorPF2e extends DetectionModeTremor {
     }
 }
 
+class ThoughtsDetectionMode extends DetectionMode {
+    constructor() {
+        super({
+            id: "senseThoughts",
+            label: "PF2E.Actor.Creature.Sense.Type.Thoughtsense",
+            type: DetectionMode.DETECTION_TYPES.OTHER,
+            walls: false,
+            angle: false,
+        });
+    }
+
+    static override getDetectionFilter(): OutlineOverlayFilter {
+        const filter = (this._detectionFilter ??= OutlineOverlayFilter.create({
+            wave: true,
+            knockout: false,
+            outlineColor: [0, 1, 0, 1],
+        }));
+        filter.thickness = 1;
+        return filter;
+    }
+
+    protected override _canDetect(visionSource: PointVisionSourcePF2e, target: PlaceableObject): boolean {
+        // Not if the target isn't a token
+        if (!(target instanceof TokenPF2e)) return false;
+
+        // Not if the token is GM-hidden
+        if (target.document.hidden) return false;
+
+        // Detection only works on creatures
+        if (!target.actor.isOfType("character") && !target.actor.isOfType("npc") && !target.actor.isOfType("familiar"))
+            return false;
+
+        // Detection cails on mindless creatures
+        if (target.actor.system?.traits?.value?.includes("mindless")) return false;
```

Comment:
```suggestion
        if (target.actor.system.traits.value.includes("mindless")) return false;
```

This also needs to be blocked by mental immunity.

Reply:
Test for `target.actor.attributes.immunities.includes("mental")` added